File: crates/generate/ess-synth/tests/fixtures/selection-prepared.go
package system_test

import (
	"example.invalid/selection/system"
	"example.invalid/selection/types/core"
	"testing"
)

func TestHostPreparesOnceThenGeneratedCodeSelects(t *testing.T) {
	event := core.Arrived{Data: "host-owned raw representation"}
	calls := 0
	prepare := func(_ string) []*core.Leg {
		calls++
		var external core.Domain = core.DomainExternal{}
		var internal core.Domain = core.DomainInternal{}
		return []*core.Leg{{Id: "external", From: "remote", Domain: &external}, {Id: "agent", From: "remote", Domain: &internal}}
	}
	prepared := prepare(event.Data)
	result, failure := system.ChooseFromPrepared(event, prepared)
	if failure != nil {
		t.Fatal(failure)
	}
	if calls != 1 || result.AgentId == nil || *result.AgentId != "agent" || result.ExternalId == nil || *result.ExternalId != "external" {
		t.Fatalf("wrong preparation/selection result: %d %+v", calls, result)
	}
}

File: crates/generate/ess-synth/tests/fixtures/selection-prepared.rs
use selection_system::choose_from_prepared;
use selection_types::core::{Arrived, Domain, Leg};
#[test]
fn host_prepares_once_then_generated_code_selects() {
    let event = Arrived { data: "host-owned raw representation".into() };
    let calls = std::cell::Cell::new(0);
    let prepare = |_: &str| { calls.set(calls.get()+1); vec![
        Some(Leg { id: "external".into(), from: "remote".into(), domain: Some(Domain::External), source: None }),
        Some(Leg { id: "agent".into(), from: "remote".into(), domain: Some(Domain::Internal), source: None }),
    ] };
    let prepared = prepare(&event.data);
    let result = choose_from_prepared(&event, &prepared).unwrap();
    assert_eq!(calls.get(), 1);
    assert_eq!(result.agent_id.as_deref(), Some("agent"));
    assert_eq!(result.external_id.as_deref(), Some("external"));
}
